### PR TITLE
[Small Feature] Add hash functions for kernel objects

### PR DESCRIPTION
**Pull request by sgiraudot**
## Rationale

With CGAL supporting C++11/14, users should be able to use some CGAL Kernel objects with `std::unordered_set` and `std::unordered_map` without too much trouble (points, segments, etc.).

All these objects already have `operator==()`, but to be usable with the unordered structure, they also need to come with a specialization of `std::hash`: this is what this PR does.

## Summary of API changes

The following Kernel objects are given a partial specialization of `std::hash`, in the cases where `Kernel::FT` is a floating-point type:

- `Aff_transformation_2.h`
- `Aff_transformation_3.h`
- `Bbox_2.h`
- `Bbox_3.h`
- `Circle_2.h`
- `Iso_cuboid_3.h`
- `Iso_rectangle_2.h`
- `Point_2.h`
- `Point_3.h`
- `Segment_2.h`
- `Segment_3.h`
- `Sphere_3.h`
- `Vector_2.h`
- `Vector_3.h`
- `Weighted_point_2.h`
- `Weighted_point_3.h`

Objects such as lines, planes or triangles are not handled because comparing the values of their internal numbers is not enough to test equality, which means that it's complicated to create a hash function that ensures that two equal objects are given the same hash.

The following line is added to the documentation of all the previous classes:

```
\cgalModels `Hashable` if `Kernel` is a cartesian kernel and if `Kernel::FT` is `Hashable`
```

With the exception of `Bbox_2` and `Bbox_3` which are simply documented as:

```
\cgalModels `Hashable`
```

as they are not templated by a Kernel and are always hashable.

## License and copyright ownership

LGPL GeometryFactory

## CHANGES.md

```markdown
### 2D and 3D Linear Geometry Kernel
 - The following objects are now hashable and thus trivially usable
   with `std::unordered_set` and `std::unordered_map`:
   `CGAL::Aff_transformation_2`, `CGAL::Aff_transformation_3`,
   `CGAL::Bbox_2`, `CGAL::Bbox_3`, `CGAL::Circle_2`,
   `CGAL::Iso_cuboid_3`, `CGAL::Iso_rectangle_2`, `CGAL::Point_2`,
   `CGAL::Point_3`, `CGAL::Segment_2`, `CGAL::Segment_3`,
   `CGAL::Sphere_3`, `CGAL::Vector_2`, `CGAL::Vector_3`,
   `CGAL::Weighted_point_2` and `CGAL::Weighted_point_3`.
```

## Submission

* [__Updated Reference Manual__](https://cgal.geometryfactory.com/~sgiraudot/Kernel_hash_functions/Kernel_23/group__PkgKernel23Ref.html)

## Status

- Developed and locally tested (GNU/Linux)
- [Small Feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Add_hash_functions_for_kernel_objects)
- pre-approuved by @lrineau 2019/06/04

Partial fix of #1116.

### CODE REVIEW

**Comment by lrineau** [comment]
For the documentation, you can document
> `Point_3<K>` is a model of [`Hashable`] if `K::FT` is a floating-point number.

[`Hashable`]: https://doc.cgal.org/latest/STL_Extension/classHashable.html

**Comment by sgiraudot** [comment]
I've completed the work on other objects, created the small feature and updated `CHANGES.md`.

I did not judge it necessary to upload the doc, as the changes are quite trivial (just adding a `\cgalModels` to several objects), but let me know if you want me to do it.

File: Installation/CHANGES.md [hunk 1]
```diff
@@ -45,6 +45,16 @@ Release date: September 2019
...
+ - The following objects are now hashable and thus trivially usable
+   with `std::unordered_set` and `std::unordered_map`:
+   `CGAL::Aff_transformation_2.h`, `CGAL::Aff_transformation_3.h`,
+   `CGAL::Bbox_2.h`, `CGAL::Bbox_3.h`, `CGAL::Circle_2.h`,
+   `CGAL::Direction_2.h`, `CGAL::Direction_3.h`,
+   `CGAL::Iso_cuboid_3.h`, `CGAL::Iso_rectangle_2.h`,
+   `CGAL::Point_2.h`, `CGAL::Point_3.h`, `CGAL::Ray_2.h`,
+   `CGAL::Ray_3.h`, `CGAL::Segment_2.h`, `CGAL::Segment_3.h`,
+   `CGAL::Vector_2.h`, `CGAL::Vector_3.h`, `CGAL::Weighted_point_2.h`
+   and `CGAL::Weighted_point_3.h`.
```
Comment (lrineau) [inline Installation/CHANGES.md]: I think you did not want those `.h` extensions after the object types.

**Comment by lrineau** [comment]
[**@sgiraudot**](https://github.com/sgiraudot) commented on [Jun 4, 2019, 10:18 AM GMT+2](https://github.com/CGAL/cgal/pull/3973#issuecomment-498571255 "2019-06-04T08:18:00Z - Replied by Github Reply Comments"):
> I've completed the work on other objects, created the small feature and updated `CHANGES.md`.

Please provide the link to the small feature. I cannot see it.

> I did not judge it necessary to upload the built doc, as the changes are quite trivial (just adding a `\cgalModels` to several objects), but let me know if you want me to do it.

Our testsuite does not check the compilation of the documentation (it does, but Doxygen rarely report errors), and that is really easy to mess the documentation with a Doxygen syntax error. Please provide a built version of the documentation, so that anybody can visually checks that everything is OK.

File: Installation/CHANGES.md (hunk 1 above)
Comment (sgiraudot) [reply Installation/CHANGES.md]: Oops, yes, that was a badly processed copy-paste... I'll correct it.

**Comment by sgiraudot** [comment]
Ok, I've updated the doc and added the link to the small feature.

**Comment by mglisse** [comment]
> in the cases where `Kernel::FT` is a floating-point type

why not in the cases where `Kernel::FT` is hashable?

**Comment by mglisse** [comment]
It seems that the `Direction_2` given by a vector and by twice that vector are the same for `==` but have different hashes?

**Comment by lrineau** [comment]
[**@mglisse**](https://github.com/mglisse) commented on [Jun 4, 2019, 1:02 PM GMT+2](https://github.com/CGAL/cgal/pull/3973#issuecomment-498624914 "2019-06-04T11:02:31Z - Replied by Github Reply Comments"):
> It seems that the `Direction_2` given by a vector and by twice that vector are the same for `==` but have different hashes?

I agree with Marc about `Direction_2` and `Direction_3`: their `operator==` is non-trivial, and those types should not be hashable. And, anyway, who wants to bet on the usefulness of hashing directions?...

**Comment by sgiraudot** [comment]
Yes, you are right about directions, I did not notice that. I'll remove them.

About the `Kernel::FT` being hashable, it's a bit more complicated, because the requirement is also that the hash does not change over time: as the representation of exact numbers may change if computations are performed on them, it seems non-trivial to create a consistent hash from them, unless I'm mistaken (if there is a way to do it, then we should include it).

**Comment by sgiraudot** [comment]
I'll also remove rays which depend on direction…

**Comment by lrineau** [comment]
We should also restrict the hashes to object of Cartesian kernels, if possible.

**Comment by lrineau** [comment]
[**@sgiraudot**](https://github.com/sgiraudot) commented on [Jun 4, 2019, 1:17 PM GMT+2](https://github.com/CGAL/cgal/pull/3973#issuecomment-498629207 "2019-06-04T11:17:39Z - Replied by Github Reply Comments"):
> About the `Kernel::FT` being hashable, it's a bit more complicated, because the requirement is also that the hash does not change over time: as the representation of exact numbers may change if computations are performed on them, it seems non-trivial to create a consistent hash from them, unless I'm mistaken (if there is a way to do it, then we should include it).

Is not that a moral contract on hashes that hashes are consistent with the value-semantic of objects? And in particular
  - if `a==b`, then `hash(a)==hash(b)`.
  - and thus hashes of const objects should not change.

**Comment by sgiraudot** [comment]
Yes, exactly, see [the doc of `std::hash`](https://en.cppreference.com/w/cpp/utility/hash):

> 4. For two parameters k1 and k2 that are equal, std::hash<Key>()(k1) == std::hash<Key>()(k2).

**Comment by mglisse** [comment]
> the requirement is also that the hash does not change over time: as the representation of exact numbers may change if computations are performed on them,

Then that number type should not be hashable? Gmpq or mpq_class seem very hashable to me, although the hash function should be provided by GMP and isn't :disappointed:. Hopefully `mpq_rational` is hashable? `Lazy_exact_nt` can only be sensibly hashed by hashing the result of `exact()`, so it is debatable if the hashing is useful, but at least it seems possible. What did you have in mind?

**Comment by afabri** [comment]
Note that `Rational` has the same problem as `Direction`.

**Comment by sgiraudot** [comment]
I have a very loose knowledge of exact numbers, so I think I was mainly thinking about `Lazy_exact_nt`. I have no idea if the other types you mention are easily hashable or not (not to mention how to do it), I only took care of the `double`/`float` case as it was the only case I felt confident in what I was doing.

**Comment by lrineau** [comment]
[**@sgiraudot**](https://github.com/sgiraudot) commented on [Jun 4, 2019, 1:40 PM GMT+2](https://github.com/CGAL/cgal/pull/3973#issuecomment-498635784 "2019-06-04T11:40:45Z - Replied by Github Reply Comments"):
> I have a very loose knowledge of exact numbers, so I think I was mainly thinking about `Lazy_exact_nt`. I have no idea if the other types you mention are easily hashable or not (not to mention how to do it), I only took care of the `double`/`float` case as it was the only case I felt confident in what I was doing.

The issue https://github.com/CGAL/cgal/issues/1116 has been stalled for three years, because there was that discussion about hashing number types. The proposal of Simon is sound: provide the hash functions of number types that are know to have hashes (actually, the hashing of a `double` is not completely trivial, because `-0==0` whereas they have two distinct bits representation, but `std::hash` deals with that for us).

I think we can discuss about extending the proposal of Simon, if that is easy, but that should not be a reason to block the integration of the version limited to `float`/`double`..

**Comment by mglisse** [comment]
> Note that `Rational` has the same problem as `Direction`.

i.e. they need to be normalized for the hash to make sense. Good thing that mpq_t are always normalized. Quotient isn't, which is why I was saying that it requires a call to `normalize()` in the hasher (or a promise from the user that they did normalize all their numbers first).

**Comment by mglisse** [comment]
> provide the hash functions of number types that are know to have hashes

That's pretty much what I was suggesting, except that I was thinking of making it automatic instead of hard-coding some types. In some sense I wanted to make it **more** independent of the problem of hashing numbers. I didn't ask to reject or block this PR.

**Comment by lrineau** [comment]
How can it be automatic? No `enable_if` and we trust that if `std::hash<FT>` compiles then it is good?

As I said, I would like to `enable_if` the `Rep_tag` of the kernel is `Cartesian_tag`. I do not think we can hash correctly homogeneous types, because of quotient by `hw`.

File: Kernel_23/include/CGAL/Kernel/hash_functions.h [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+#define CGAL_KERNEL_HASH_FUNCTIONS_H
+
+namespace CGAL
+{
+
+template <typename K>
+inline std::enable_if_t<std::is_floating_point<typename K::FT>::value, std::size_t>
+hash_value (const Aff_transformation_2<K>& transform)
+{
+  std::size_t result = boost::hash_value(transform.cartesian(0,0));
```
Comment (lrineau) [inline Kernel_23/include/CGAL/Kernel/hash_functions.h]: `hash_value` must be called by ADL, without the namespace. Or, if you want to specify a namespace, it must be `CGAL`.

Travis tests are fully red: https://travis-ci.org/CGAL/cgal/builds/541190492. It probably does not compile on your machine either.

Comment (sgiraudot) [reply Kernel_23/include/CGAL/Kernel/hash_functions.h]: It does compile using Clang (I just double-checked), but fails with GCC.

But in that case, `hash_value` is applied on a double/float type, so it does come from Boost, doesn't it? I pretty much copied what I did for [`CGAL::grid_simplify_points()`](https://github.com/CGAL/cgal/blob/master/Point_set_processing_3/include/CGAL/grid_simplify_point_set.h#L80) a few years ago.

**Comment by mglisse** [comment]
> How can it be automatic? No `enable_if` and we trust that if `std::hash<FT>` compiles then it is good?

Sure, why not? Unless I am missing something, the main reason not to would be so users can SFINAE-detect if Point_2 is hashable, and if that seems important, we could enable_if it.

File: Kernel_23/include/CGAL/Kernel/hash_functions.h (hunk 2 above)
Comment (lrineau) [reply Kernel_23/include/CGAL/Kernel/hash_functions.h]: Actually my suggestion was wrong. The correct one is:
  - call by ADL (without the namespace),
  - but add a `using boost::hash_value;` before, for the cases where the FT does not have a namespace, like `double`.

Maybe the errors in Travis are just about a missing `#include` for `boost::hash_value`. But in general `hash_value` must be called by ADL (with the additional `using`).

**Comment by sgiraudot** [comment]
Alright, so I fixed the issues and after discussion with Laurent, we changed the condition: now the kernel needs to be cartesian and the FT just needs to be hashable (which means if the FT of Epeck becomes hashable, the objects of Epeck will be usable with unordered set/map without changing anything).

**Comment by sgiraudot** [comment]
Just for information, I tested the following:
- Using `Simple_homogeneous` returns an error that says that no `hash_value` can be found for the wanted object (`Point_2`, etc.)
- Using `Exact_predicates_exact_constructions` returns an error that says that no `std::hash<FT>` can be found.

So I think these are pretty clear errors and that match the documentation quite well, which should prevent any bad surprise for the user.

**Comment by maxGimeno** [comment]
There are many errors in Travis, I cannot add this PR into integration until it is all clear.

**Comment by lrineau** [comment]
The functions `hash_value` for Bbox_23 are not template functions, and must be declared `inline`.

**Comment by lrineau** [comment]
Tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-62.shtml

**Pull request merged**